lrcget: init at 0.4.0

0x61nas opened the pull request:
## Description of changes

Create `lrcget` package, requested in #312996

### Try it
```sh
nix run github:anas-contribs/nixpkgs/lrcget#lrcget

```

- Closes: #312996

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [ ] [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,99 @@
+    '';
+  };
+in
+rustPlatform.buildRustPackage rec {
```
is rec needed?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4065

Scrumplex commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,99 @@
+    rev = "${version}";
+    hash = "sha256-9Wtb8Y+cj1ec8vNxvvs3wso62TavNveHRh2Gfx7Br4w=";
+  };
+  frontendBuild = buildNpmPackage {
+    inherit version src;
+    pname = "lrcget-ui";
+    # FIXME: This is a workaround, because we have an git dependency node_modules/lrc-kit contains install scripts
+    # but has no lockfile, which is something that will probably break.
+    forceGitDeps = true;
+    distPhase = "true";
+    dontInstall = true;
+    # To fix `npm ERR! Your cache folder contains root-owned files`
+    makeCacheWritable = true;
+
+    npmDepsHash = "sha256-qQ5UMO3UuD6IvUveTRF35qTlGq5PMbxp1Q4UroDqVtk=";
+
+    # packageJSON = ./package.json;
+    postBuild = ''
+      cp -r dist/ $out
+    '';
+  };
```
I think you could add this as its own attribute to the package.

i.e.

```nix
rustPlatform.buildRustPackage rec {
  inherit version src pname;

  frontend = buildNpmPackage {
		...
  };
  ...
}
```
This would also allow you to reference the output of this in `postBuild` using `$frontend`. The main reason for doing this would be to allow overriding this in case someone needs to for whatever reason, as well as allowing users to build the frontend (and test the package more easily)

0x61nas replied:
>  The main reason for doing this would be to allow overriding this in case someone needs to for whatever reason, as well as allowing users to build the frontend (and test the package more easily)

if you have time, i would love to see an example for this, ama new in nix

Scrumplex commented:
Result of `nixpkgs-review pr 318068` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lrcget</li>
  </ul>
</details>

Aleksanaa commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,100 @@
+  };
+in
+rustPlatform.buildRustPackage rec {
+  inherit version src pname;
```
No need to put it outside, since you've already used `rec`

Aleksanaa commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,100 @@
+  doCheck = false;
+
+  postInstall = ''
+    install -DT icons/128x128@2x.png $out/share/icons/hicolor/128x128@2/apps/lrcget.png
```
Do we have `@2` in the freedesktop specification? I'm not sure

0x61nas replied:
am not sure, but `squirreldisk` (the package that I've taken as reference) did that, ref: https://github.com/NixOS/nixpkgs/blob/nixos-24.05/pkgs/by-name/sq/squirreldisk/package.nix#L77

Scrumplex reviewed (commented):
I am currently on vacation but I forgot to send these comments

0x61nas commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,100 @@
+  # Also modify tauri.conf.json so that it expects the resources at the new location
+  postPatch = ''
+    mkdir -p frontend
+    cp -r ${frontend}/* frontend
```
in my opinion the `{}` make it easier read

Scrumplex replied:
The `${var}` syntax is semantically different to `$var`

`${var}` means that Nix will substitute the variable during evaluation time and as you use `rec` here it will always be the attribute in this file.

`$var` is just the notation to let Bash substitute the variable, so its scope won't be the this nix file, but rather the (environment) variables set by stdenv.

The reason this is important is that `$var` will always have the actual final value of `var` while `${var}` (assuming rec is used) will only have the initial value.

0x61nas replied:
ah, that's new to me, thanks for clarifying

AndersonTorres commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,96 @@
+  frontend = buildNpmPackage {
+    inherit version src;
+    pname = "lrcget-ui";
+    # FIXME: This is a workaround, because we have an git dependency node_modules/lrc-kit contains install scripts
```
```suggestion
    # FIXME: This is a workaround, because we have a git dependency node_modules/lrc-kit contains install scripts
```

AndersonTorres commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,96 @@
+
+    npmDepsHash = "sha256-qQ5UMO3UuD6IvUveTRF35qTlGq5PMbxp1Q4UroDqVtk=";
+
+    # packageJSON = ./package.json;
```
```suggestion
```

AndersonTorres commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,96 @@
+  '';
+
+  nativeBuildInputs = [ pkg-config wrapGAppsHook3 copyDesktopItems ];
+  buildInputs = [ dbus openssl gtk3 webkitgtk alsa-lib ];
```
```suggestion

  buildInputs = [ dbus openssl gtk3 webkitgtk alsa-lib ];
```

IldenH commented:
0.4 is out
https://github.com/tranxuanthang/lrcget/releases/tag/0.4.0.

Scrumplex commented:
Result of `nixpkgs-review pr 318068` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lrcget</li>
  </ul>
</details>

Scrumplex commented:
I would love to co-maintain this :D

If you don't mind, please add `Scrumplex` to `meta.maintainers`

0x61nas commented:
sure @Scrumplex, thx for your interest

0x61nas committed:
lrcget: init at 0.4.0

0x61nas committed:
lrcget: add maintainer Scrumplex

Scrumplex has volunteered to be a co-maintainer in https://github.com/NixOS/nixpkgs/pull/318068#issuecomment-2268520944

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1896

gador commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,110 @@
+  stdenv,
+  copyDesktopItems,
+  makeDesktopItem,
+  alsa-lib,
```
```suggestion
  alsa-lib,
  darwin,
```

gador commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,110 @@
+  nativeBuildInputs = [
+    pkg-config
+    wrapGAppsHook3
+    copyDesktopItems
```
```suggestion
    copyDesktopItems
    rustPlatform.bindgenHook
```

gador commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,110 @@
+    copyDesktopItems
+  ];
+
+  buildInputs = [
+    dbus
+    openssl
+    gtk3
+    webkitgtk
+    alsa-lib
+  ];
```
```suggestion
  buildInputs =
    [
      dbus
      openssl
      gtk3
    ]
    ++ lib.optionals (!stdenv.isDarwin) [
      webkitgtk
      alsa-lib
    ]
    ++ lib.optionals stdenv.isDarwin [
      darwin.apple_sdk.frameworks.CoreAudio
      darwin.apple_sdk.frameworks.WebKit
    ];
  ];
```

gador commented on pkgs/by-name/lr/lrcget/package.nix:
```diff
@@ -0,0 +1,110 @@
+    ];
+    mainProgram = "lrcget";
+    platforms = with lib.platforms; unix ++ windows;
+    broken = stdenv.isDarwin; # needing some apple_sdk packages
```
```suggestion
```

gador reviewed (commented):
Adding the required frameworks enables the build on `aarch64-darwin`. I haven't tested the app yet (only that it compiles), because I don't have access to my darwin desktop until later today. 

EDIT: Tested on `aarch64-darwin` and works as expected
